test/unit/test_data_utils.py

```diff
@@ -79,6 +88,34 @@ def test_parse_csv_dmatrix(self):
 
                 self.assertEqual(no_weight_test_features, single_node_dmatrix.feature_names)
 
+    def test_parse_csv_dmatrix_pipe(self):
```

Comment:
The code for testing pipe mode has a lot of shared logic, can we abstract that out to a helper method here

Reply:
Will do.